### ISSUE TITLE
[Camel 3.5/3.4.3] xml-io should pass namespace info to NamespaceAware elements

### ISSUE THREAD

**Issue by kesavkolla** [apache/camel-quarkus#1497]
Hi I'm trying to use xml based route using routes. I've a xpath expression which uses namespace and it's not taking defined namespaces in routes xml. How do I define namespaces? Here is my route xml

```xml
<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns="http://camel.apache.org/schema/spring"
        xmlns:soap-env="http://www.w3.org/2003/05/soap-envelope"
        xsi:schemaLocation="
            http://camel.apache.org/schema/spring
            http://camel.apache.org/schema/spring/camel-spring-3.4.2.xsd">
    <route id="myroute">
        <from uri="platform-http:/handler?httpMethodRestrict=POST"/>
        <setProperty name="MessageType">
            <xpath resultType="java.lang.String" logNamespaces="true">
                local-name(/soap-env:Envelope/soap-env:Body/*[1])
            </xpath>
        </setProperty>
        <log message="${exchangeProperty.MessageType}"/>
        <setBody>
            <simple>
                Hello ${body}
            </simple>
        </setBody>
    </route>
</routes>
```

When I try to post XML file to handler endpoint I get the following error:

```
org.apache.camel.language.xpath.InvalidXPathException: Invalid xpath: local-name(/soap-env:Envelope/soap-env:Body/*[1]). Reason: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Prefix must resolve to a namespace: soap-env
    at org.apache.camel.language.xpath.XPathBuilder.evaluateAs(XPathBuilder.java:943)
    at org.apache.camel.language.xpath.XPathBuilder.evaluate(XPathBuilder.java:921)
    at org.apache.camel.language.xpath.XPathBuilder.evaluate(XPathBuilder.java:240)
    at org.apache.camel.processor.SetPropertyProcessor.process(SetPropertyProcessor.java:47)
    at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:395)
    at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:148)
    at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
    at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
    at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:286)
    at org.apache.camel.quarkus.component.platform.http.runtime.QuarkusPlatformHttpConsumer.lambda$doStart$2(QuarkusPlatformHttpConsumer.java:123)
.....................
Caused by: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Prefix must resolve to a namespace: soap-env
    at java.xml/com.sun.org.apache.xpath.internal.jaxp.XPathImpl.compile(XPathImpl.java:170)
    at org.apache.camel.language.xpath.XPathBuilder.createXPathExpression(XPathBuilder.java:1170)
    at org.apache.camel.language.xpath.XPathBuilder.evaluateAs(XPathBuilder.java:941)
    ... 29 more
Caused by: javax.xml.transform.TransformerException: Prefix must resolve to a namespace: soap-env
    at java.xml/com.sun.org.apache.xpath.internal.compiler.XPathParser.error(XPathParser.java:621)
    at java.xml/com.sun.org.apache.xpath.internal.compiler.Lexer.mapNSTokens(Lexer.java:637)
    at java.xml/com.sun.org.apache.xpath.internal.compiler.Lexer.tokenize(Lexer.java:269)
```

Can anyone point me how to define custom namespaces?

**Comment by ppalaga** [comment]
Rewriting the xpath expression (so that it ignores any namespace info) could perhaps serve as a workaround:

```xml
<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns="http://camel.apache.org/schema/spring"
        xsi:schemaLocation="
            http://camel.apache.org/schema/spring
            http://camel.apache.org/schema/spring/camel-spring-3.4.2.xsd">
    <route id="myroute">
        <from uri="platform-http:/handler?httpMethodRestrict=POST"/>
        <setProperty name="MessageType">
            <xpath resultType="java.lang.String" logNamespaces="true">
                local-name(/*[local-name() = 'Envelope'/*[local-name() = 'Body']/*[1])
            </xpath>
        </setProperty>
        <log message="${exchangeProperty.MessageType}"/>
        <setBody>
            <simple>
                Hello ${body}
            </simple>
        </setBody>
    </route>
</routes>
```

**Comment by kesavkolla** [comment]
:smiley: Yeah that is a workaround. This will cause XPath expression a bit complex. Specially when we have to deal with nested paths.

**Comment by ppalaga** [comment]
Yeah, I am not doubting there should be some user friendly way to set the prefix -> namespace mapping. The workaround may help in some situations while we do not have a proper fix.

**Comment by ppalaga** [comment]
It looks like the Jaxb based XMLRoutesDefinitionLoader dropped in https://github.com/apache/camel-quarkus/pull/1292 would work as expected. The `xml-io` XMLRoutesDefinitionLoader (the only one we offer ATM) ignores namespaces. I think it needs to get fixed in Camel. Let me check that.

**Comment by ppalaga** [comment]
> It looks like the Jaxb based XMLRoutesDefinitionLoader dropped in #1292 would work as expected.

In other words, using camel-quarkus 1.0.0-CR2 and replacing the `camel-quarkus-xml-io` dependency with `camel-quarkus-xml-jaxb` might be another workaround.

**Comment by ppalaga** [comment]
Blocked by https://github.com/apache/camel/pull/4053

**Comment by ppalaga** [comment]
Thanks for reporting @kesavkolla !

**Comment by kesavkolla** [comment]
If I remove camel-quarkus-xml-io with camel-quarkus-xml-jaxb I get the following exception:

```java
2020-07-28 08:49:55,033 ERROR [org.apa.cam.qua.mai.CamelMainRuntime] (main) Failed to start application: java.lang.UnsupportedOperationException: Please add a dependency to camel-quarkus-xml-io
	at org.apache.camel.quarkus.core.DisabledXMLRoutesDefinitionLoader.loadRoutesDefinition(DisabledXMLRoutesDefinitionLoader.java:28)
	at org.apache.camel.quarkus.main.CamelMainRoutesCollector.collectXmlRoutesFromDirectory(CamelMainRoutesCollector.java:74)
	at org.apache.camel.main.RoutesConfigurer.configureRoutes(RoutesConfigurer.java:97)
	at org.apache.camel.main.BaseMainSupport.configureRoutes(BaseMainSupport.java:556)
	at org.apache.camel.main.BaseMainSupport.postProcessCamelContext(BaseMainSupport.java:576)
	at org.apache.camel.quarkus.main.CamelMain.initCamelContext(CamelMain.java:91)
	at org.apache.camel.quarkus.main.CamelMain.doInit(CamelMain.java:61)
	at org.apache.camel.support.service.BaseService.init(BaseService.java:83)
	at org.apache.camel.quarkus.main.CamelMain.startEngine(CamelMain.java:112)
	at org.apache.camel.quarkus.main.CamelMainRuntime.start(CamelMainRuntime.java:46)
	at org.apache.camel.quarkus.core.CamelBootstrapRecorder.start(CamelBootstrapRecorder.java:45)
	at io.quarkus.deployment.steps.CamelBootstrapProcessor$boot-173480958.deploy_0(CamelBootstrapProcessor$boot-173480958.zig:97)
	at io.quarkus.deployment.steps.CamelBootstrapProcessor$boot-173480958.deploy(CamelBootstrapProcessor$boot-173480958.zig:36)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:644)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:91)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)

```

**Comment by ppalaga** [comment]
You need to take care that you use camel-quarkus 1.0.0-CR2 (rather than 1.0.0-CR3 that throws the above exception). If you generated your pom.xml using https://code.quarkus.io , you need to downgrade `quarkus.platform.version` to 1.5.0.Final (which should transitively pull camel-quarkus-bom 1.0.0-CR2, IIRC).

**Comment by kesavkolla** [comment]
Is this issue fixed in latest 1.0.0 version?

**Comment by ppalaga** [comment]
No, the fix was done in the upcoming Camel 3.5.0 and also in the recently released Camel 3.4.3. So you'll have to wait for Camel Quarkus 1.1.0 or 1.0.1. 
1.0.1 may appear within 10 days or so.
